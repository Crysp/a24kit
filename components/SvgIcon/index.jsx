import React from 'react';
import PropTypes from 'prop-types';
import Wrapper from './styled';


/**
 * Обертка для любой svg иконки
 * @version 1.0.0
 */
export const sizes = {
    AUTO: 'auto',
    SIZE_8: 8,
    SIZE_12: 12,
    SIZE_16: 16,
    SIZE_20: 20,
    SIZE_24: 24,
    SIZE_32: 32,
};
export default class SvgIcon extends React.PureComponent {
    static propTypes = {
        /**
         * SVG разметка
         */
        children: PropTypes.node.isRequired,
        /**
         * Определяет видимую область
         * ([стандартный атрибут SVG](https://developer.mozilla.org/ru/docs/Web/SVG/Attribute/viewBox))
         */
        viewBox: PropTypes.string.isRequired,
        /**
         * Размер иконки
         * - auto (заполняет все пространство контейнера)
         * - 8, 12, 16, 20, 24, 32 (наиболее используемые размеры)
         *
         * Исключения задаются через CSS стили
         */
        size: PropTypes.oneOf([
            sizes.AUTO,
            sizes.SIZE_8,
            sizes.SIZE_12,
            sizes.SIZE_16,
            sizes.SIZE_20,
            sizes.SIZE_24,
            sizes.SIZE_32,
        ]),
        /**
         * Кастомные стили
         */
        style: PropTypes.shape(),
    };

    static defaultProps = {
        style: {},
        size: sizes.AUTO,
    };

    render() {
        const {
            children,
            ...other
        } = this.props;
        return (
            <Wrapper
                focusable="false"
                aria-hidden="true"
                {...other}
            >
                {children}
            </Wrapper>
        );
    }
}
