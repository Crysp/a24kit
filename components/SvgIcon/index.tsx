import React, { SVGAttributes } from 'react';
import Wrapper from './styled';

export interface Props extends SVGAttributes<SVGElement> {
    children: React.ReactNode
    size?: string | number
    style?: object
}

export const sizes = {
    AUTO: 'auto',
    SIZE_8: 8,
    SIZE_12: 12,
    SIZE_16: 16,
    SIZE_20: 20,
    SIZE_24: 24,
    SIZE_32: 32,
};
export default class SvgIcon extends React.PureComponent<Props> {
    static defaultProps = {
        style: {},
        size: sizes.AUTO,
    };

    render() {
        const {
            children,
            ...other
        } = this.props;
        return (
            <Wrapper
                focusable="false"
                aria-hidden="true"
                {...other}
            >
                {children}
            </Wrapper>
        );
    }
}
