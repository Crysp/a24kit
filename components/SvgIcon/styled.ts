import styled, { css } from 'styled-components';

const getSize = (props) => {
    if (typeof props.size === 'number') {
        return css`
            width: ${props.size}px;
            height: ${props.size}px;
        `;
    }
    return css`
        width: 100%;
        height: 100%;
    `;
};

export default styled.svg`
    display: inline-block;
    fill: currentColor;
    stroke: currentColor;
    stroke-width: 0;
    user-select: none;
    flex-shrink: 0;
    ${getSize}
`;
