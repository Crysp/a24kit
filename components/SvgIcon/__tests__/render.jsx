import React from 'react';
import { create } from 'react-test-renderer';
import SvgIcon, { sizes } from '../index';


describe('Верстка компонента <SvgIcon>', () => {
    it('Дефолтная', () => {
        const tree = create(<SvgIcon viewBox="0 0 100 100"><rect width="100" height="100" /></SvgIcon>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером с родителя', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.AUTO}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером 8x8', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.SIZE_8}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером 12x12', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.SIZE_12}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером 16x16', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.SIZE_16}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером 24x24', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.SIZE_24}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('Размером 32x32', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" size={sizes.SIZE_32}><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('С дополнительным CSS классом', () => {
        const icon = <SvgIcon viewBox="0 0 100 100" className="custom"><rect width="100" height="100" /></SvgIcon>;
        const tree = create(icon).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
