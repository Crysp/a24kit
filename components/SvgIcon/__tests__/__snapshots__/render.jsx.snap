// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Верстка компонента <SvgIcon> Дефолтная 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa dTKkTG"
  focusable="false"
  size="auto"
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером 8x8 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa eRdBcm"
  focusable="false"
  size={8}
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером 12x12 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa clsejH"
  focusable="false"
  size={12}
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером 16x16 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa hDLkZN"
  focusable="false"
  size={16}
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером 24x24 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa ccArRa"
  focusable="false"
  size={24}
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером 32x32 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa iWSSjF"
  focusable="false"
  size={32}
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> Размером с родителя 1`] = `
<svg
  aria-hidden="true"
  className="sc-bdVaJa dTKkTG"
  focusable="false"
  size="auto"
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;

exports[`Верстка компонента <SvgIcon> С дополнительным CSS классом 1`] = `
<svg
  aria-hidden="true"
  className="custom sc-bdVaJa dTKkTG"
  focusable="false"
  size="auto"
  style={Object {}}
  viewBox="0 0 100 100"
>
  <rect
    height="100"
    width="100"
  />
</svg>
`;
